### ISSUE TITLE
Does not consider the temperature of the AMD GPU

### ISSUE THREAD

**Issue by CaptainMorgan12** [GnomedDev/T2FanRD#3]
Environment: MacPro 7,1 , 2022.140.5.0.0 (iBridge: 21.16.6074.0.0,0), Linux 6.10.9-1-t2-noble, wayland, Ubuntu 24.04.1 LTS, AMD Radeon™ Pro 580X, Mesa LTS 24.2.2~kisak1~n, Reported Sep 9th 2024.

Situation: Play steam rebel galaxy outlaw watch the "edge amdgpu-pci-0700" temperature of the GPU go up to 100degC and observe the fan does not kick on. The CPU temperature during game play stays at around 40degC since it isn't used.

![image](https://github.com/user-attachments/assets/a8e1254a-896f-4e10-857e-a0e75c28127d)

**Comment by GnomedDev** [comment]
I do not have a machine to test this out with, so cannot implement this, happy to accept a PR though.

**Comment by CaptainMorgan12** [comment]
One other thing i notice that may or maynot be related, but changing the fan 1 config values:

[Fan1]
low_temp=1
high_temp=20
speed_curve=linear
always_full_speed=true

or any other values has no effect, it basically just runs at 500rpm, the other fans 2, 3, 4, 5 do change based on measured temperature. I believe the fan 1 is the one that cools the GPU in the first slot so the most important one and I can't get it to spin up past 500 :-/. Also fan 2 only sluggishly responds to any changes in the conf, the only fans that seem to follow what the user puts in is fan 3 and fan 4.

I will test any new files you can generate if you could:
1. use same configuration for temp measurement used for fan 3, and 4 on fan 1 and 2 and update program i'll test
2. adjust all fan outputs with minimum adjusted to say 700 and i can test that as well without changing the temperature sensor configuration

**Comment by patdohere** [comment]
> I do not have a machine to test this out with, so cannot implement this, happy to accept a PR though.

I have a machine with 2 AMD W6800X Duo so this would be beneficial to implement.  Looking for help on how to setup dev environment.  I was looking at this line and the difference would be to change it to look at a glob of GPU temperature files.

https://github.com/GnomedDev/T2FanRD/blob/85027878e4d7fa0170fea1213d6f8dd972d60e83/src/main.rs#L125
```
fn find_gpu_temp_file(temp_buf: &mut String) -> Result<Option<std::fs::File>> {
    let temps = glob::glob("/sys/class/drm/card0/device/hwmon/hwmon*/temp1_input")?;
    Ok(find_temp_file(temps, temp_buf))
}
```

Would need to change this to look at additional cards.  As the 2019 Mac Pro enumerates them starting with `card1` instead.  I.e. for my system the temp files would be as follows.

```
ls /sys/class/drm/card*/device/hwmon/hwmon*/temp1_input
/sys/class/drm/card1/device/hwmon/hwmon8/temp1_input  /sys/class/drm/card2/device/hwmon/hwmon9/temp1_input  /sys/class/drm/card3/device/hwmon/hwmon10/temp1_input  /sys/class/drm/card4/device/hwmon/hwmon11/temp1_input
```

My guess would be to enumerate a list of temperature files and then write a method to process them.  Probably use the highest.

**Comment by patdohere** [comment]
> > ~~One other thing i noticed that may or may not be related, but changing the fan 1 config values:
> > [Fan1]
> > low_temp=1
> > high_temp=20
> > speed_curve=linear
> > always_full_speed=true
> > or any other values than those shown has no effect, it basically just runs at 500rpm, the other fans 2, 3, 4 do change based on measured temperature. I believe the fan 1 is the one that cools the GPU in the first slot - so the most important one - and I can't get it to spin up past 500 :-/. Also fan 2 only sluggishly responds to any changes in the conf, the only fans that seem to follow what the user puts in are fan 3 and fan 4.
> > I will test any new files you can generate if you could:
> > 
> > 1. use same configuration for temp measurement used for fan 3, and 4 on fan 1 and 2 and update program i'll test
> > 2. adjust all fan outputs with minimum adjusted to say 700 and i can test that as well without changing the temperature sensor configuration~~
> 
> Ok that was user error i think by installing lm-sensors, fancontrol, macfand, mbpfan and a bunch of other programs to try to control the fans, the daemons might have been stepping on each other since most of those controlled fan 1, and/or 2, but never fan 3, 4. After purging everything F1-F4 can be controlled through T2fanrd.
> 
> So the issue boils down to the temperature sensor T2fanrd uses, which is CPU based after doing some testing. If I run Steam where CPU is cool, but GPU is only used none of the fans recognize that. So one thing one could do is use the CPU+GPU temperature average for Fan1 and Fan2 and the CPU, or average mac temp for Fan3 and Fan4. By tweaking the min/max temp using the T2fanrd configuration file i can get the Fan1 to run at say 900-1100rpm, which seems to be just enough for the GPU temperature not going critical - it hovers around 106degC at that rpm (the downside is of course that when doing CPU heavy workload Fan1 is much too loud and is cooling the GPU which may not be doing any work).

Yeah from my understanding is than the fans enumerated by T2FanRD on a 2019 Mac Pro
Fan 1 -> is the one blower exhaust fan in the back
![Image](https://github.com/user-attachments/assets/4cedb2c8-19b5-4781-b23f-cecfcd07660c)

Fan 2-4 -> are the three intake fans in the front
![Image](https://github.com/user-attachments/assets/6c636d47-ec2c-4826-bb09-08fecbd7e2c8)

**Comment by sharpenedblade** [comment]
> Looking for help on how to setup dev environment.

`git clone`, then `cargo build`. You might need to install `-devel` packages for some c libs but cargo will tell you what you are missing

> I was looking at this line and the difference would be to change it to look at a glob of GPU temperature files.

This was designed for laptops with only one gpu. To support multiple you need to change the code to return `Vec<std::fs::File>` from `find_gpu_temp_file()`. `read_temp_file()` should then pick the highest temperature out from the `Vec` to output, but this would mean you have a function for the cpu temperature and a different one for the gpu temperature.

A better approach is to make a `Vec<File>` to store all of the temperature files, including the cpu temperature. You can then pick the highest temperature in this list as the "real" temperature. All of the `cpu_temp_file`, and `gpu_temp_file` variables would replaced by a single `Vec<File>`.
